#! /bin/sh /usr/share/dpatch/dpatch-run
## 10_fix_potential_segfaults.dpatch by Stefan Fritsch <sf@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix out of bounds / NULL ptr accesses found by cppcheck

@DPATCH@
diff -urNad '--exclude=CVS' '--exclude=.svn' '--exclude=.git' '--exclude=.arch' '--exclude=.hg' '--exclude=_darcs' '--exclude=.bzr' mp3gain-1.5.1~/mpglibDBL/interface.c mp3gain-1.5.1/mpglibDBL/interface.c
--- mp3gain-1.5.1~/mpglibDBL/interface.c	2005-01-18 09:56:46.000000000 +0100
+++ mp3gain-1.5.1/mpglibDBL/interface.c	2010-04-21 00:24:02.505591419 +0200
@@ -336,7 +336,8 @@
 {
   int i,pos;
   struct buf *ibuf=mp->tail;
-  unsigned char xing[48];
+  /* GetVbrTag may read more than 48 bytes */
+  unsigned char xing[156] = {0};
   VBRTAGDATA pTagData;
 
   pos = ibuf->pos;
@@ -344,8 +345,8 @@
   for (i=0; i<bytes; ++i) {
     while(pos >= ibuf->size) {
       ibuf  = ibuf->next;
-      pos = ibuf->pos;
       if(!ibuf) 	return -1; /* fatal error */
+      pos = ibuf->pos;
     }
     ++pos;
   }
@@ -353,8 +354,8 @@
   for (i=0; i<48; ++i) {
     while(pos >= ibuf->size) {
       ibuf  = ibuf->next;
-      pos = ibuf->pos;
       if(!ibuf) 	return -1; /* fatal error */
+      pos = ibuf->pos;
     }
     xing[i] = ibuf->pnt[pos];
     ++pos;
@@ -396,11 +397,11 @@
     b[0]=b[1]; b[1]=b[2]; b[2]=b[3];
     while(pos >= buff->size) {
       buff  = buff->next;
-      pos = buff->pos;
       if(!buff) {
 	return -1;
 	/* not enough data to read 4 bytes */
       }
+      pos = buff->pos;
     }
     b[3] = buff->pnt[pos];
     ++pos;
